pkg/virtctl/adm/logverbosity/logverbosity.go

```diff
@@ -0,0 +1,406 @@
+
+	// cannot specify "reset" and "all" flag at the same time
+	cmd.MarkFlagsMutuallyExclusive("reset", allComponent)
+
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+	return cmd
+}
+
+// Command line flag syntax:
+//
+//	OK: --flag=x
+//	NG: --flag x
+//
+// Another caveat of NoOptDefVal is that
+// we cannot use the "--flag x" syntax, because "--flag x" only applies to flags without a default value.
+// See vendor/github.com/spf13/pflag/flag.go, especially note the order of the if clause below.
+// https://github.com/kubevirt/kubevirt/blob/main/vendor/github.com/spf13/pflag/flag.go#L983C1-L998C3
+func usage() string {
+	return `  # reset (to default) log-verbosity for all components:
+  {{ProgramName}} adm logVerbosity --reset
+
+  # show log-verbosity for all components:
+  {{ProgramName}} adm log-verbosity --all
+  # set log-verbosity to 3 for all components:
+  {{ProgramName}} adm log-verbosity --all=3
+
+  # show log-verbosity for virt-handler:
+  {{ProgramName}} adm log-verbosity --virt-handler
+  # set log-verbosity to 7 for virt-handler:
+  {{ProgramName}} adm log-verbosity --virt-handler=7
+
+  # show log-verbosity for virt-handler and virt-launcher:
+  {{ProgramName}} adm log-verbosity --virt-handler --virt-launcher
+  # set log-verbosity for virt-handler to 7 and virt-launcher to 3:
+  {{ProgramName}} adm log-verbosity --virt-handler=7 --virt-launcher=3
+
+  # reset all components to default besides virt-handler which is 7:
+  {{ProgramName}} adm log-verbosity --reset --virt-handler=7
+  # set all components to 3 besides virt-handler which is 7:
+  {{ProgramName}} adm log-verbosity --all=3 --virt-handler=7`
+}
+
+// component name to JSON name
+func getJSONNameByComponentName(componentName string) string {
+	var componentNameToJSONName = map[string]string{
+		"virt-api":        "virtAPI",
+		"virt-controller": "virtController",
+		"virt-handler":    "virtHandler",
+		"virt-launcher":   "virtLauncher",
+		"virt-operator":   "virtOperator",
+		allComponent:      allComponent,
+	}
+	return componentNameToJSONName[componentName]
+}
+
+func detectInstallNamespaceAndName(virtClient kubecli.KubevirtClient) (namespace, name string, err error) {
+	kvs, err := virtClient.KubeVirt(k8smetav1.NamespaceAll).List(&k8smetav1.ListOptions{})
+	if err != nil {
+		return "", "", fmt.Errorf("could not list KubeVirt CRs across all namespaces: %v", err)
+	}
+	if len(kvs.Items) == 0 {
+		return "", "", errors.New("could not detect a KubeVirt installation")
+	}
+	if len(kvs.Items) > 1 {
+		return "", "", errors.New("invalid kubevirt installation, more than one KubeVirt resource found")
+	}
+	namespace = kvs.Items[0].Namespace
+	name = kvs.Items[0].Name
+	return
+}
+
+func hasVerbosityInKV(kv *v1.KubeVirt) (map[string]uint, error) {
+	verbosityMap := map[string]uint{} // key: component name, value: verbosity
+
+	// if LogVerbosity is not present in the KubeVirt CR, skip here
+	if kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity != nil {
```

Comment:
Please, check if `DeveloperConfiguration` is nil before this condition. This is a pointer, see https://github.com/kubevirt/kubevirt/blob/main/staging/src/kubevirt.io/api/core/v1/types.go#L2319

Reply:
Sorry, but let me confirm. Are you suggesting something like this?
```
if kv.Spec.Configuration.DeveloperConfiguration != nil {
	if kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity != nil {
		...
	}
}
```
Based on my previous research below, I did not find a case where `DeveloperConfiguration` was nil (`DeveloperConfiguration` was always present).
https://github.com/kubevirt/kubevirt/pull/10244#discussion_r1372424327

On the other hand, I understand the definition to mean that it could be nil.
`DeveloperConfiguration *DeveloperConfiguration json:"developerConfiguration,omitempty".`
Also, although I have checked the value of `DeveloperConfiguration` in some cases, I am not 100% sure that we can say that `DeveloperConfiguration` is always present.

So I understand that it is safe to have the condition, and I will add the line, if my understanding of your suggestion is correct.

In that sense, do you want to add one more check if `Configuration` is nil before `DeveloperConfiguration` check?
https://github.com/kubevirt/kubevirt/blob/main/staging/src/kubevirt.io/api/core/v1/types.go#L1873

Reply:
`Configuration` can't be `nil`, it holds the type `KubeVirtConfiguration`. You could check if it is empty, but this for sure is not needed.

`DeveloperConfiguration` might not be `nil` in most scenarios as it might be initialized with an empty structure. That could change in the future and would crash virtctl. I think it is safe to add the check here.

Reply:
@hshitomi you can write something like:
```golang
if kv.Spec.Configuration.DeveloperConfiguration != nil && kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity != nil
```
Hence, no need of a nested if

Reply:
Thank you for the clarification. Will change it to that.

Now, if we need to take into account that `DeveloperConfiguration` is `nil`, we need to add more codes.

In the case of `set` operation, we need to patch and add an empty `DeveloperConfiguration` before patching and adding a `LogVerbosity` object (something like below). This is because the `add` patch will return an error if there is a missing key (`DeveloperConfiguration`) in the path to the endpoint (`LogVerbosity`).
```
		if !hasDeveloperConfiguration {
			// if DeveloperConfiguration is absent, add DeveloperConfiguration first
			addPatch(patchData, patchAdd, dcPath, &v1.DeveloperConfiguration{})
		}
		addPatch(patchData, patchAdd, lvPath, currentLv)
```
I will also add test cases for no `DeveloperConfiguration`.

Reply:
Done